library/core/src/num/ranged.rs

```diff
@@ -0,0 +1,108 @@
+//! Definitions for `Ranged` type.
+
+use crate::ops::{CoerceUnsized, DispatchFromDyn, RangeInclusive};
+
+#[cfg_attr(not(bootstrap), lang = "ranged")]
+#[derive(Copy, Clone, PartialEq, Eq, PartialOrd, Ord, Hash, Debug)]
+#[repr(transparent)]
+#[unstable(feature = "ranged_int", issue = "none")]
+/// A type that can only represent types in the given range.
+/// Layout optimizations will take.
+pub struct Ranged<T, const RANGE: RangeInclusive<u128>>(T);
+
+impl<T: CheckInRange, const RANGE: RangeInclusive<u128>> Ranged<T, RANGE> {
+    /// Create a new `Ranged` value if the passed argument is within `RANGE`.
+    #[unstable(feature = "ranged_int", issue = "none")]
+    #[inline(always)]
+    pub fn new(i: T) -> Option<Self> {
+        i.in_range(RANGE).then(|| Self(i))
+    }
+
+    /// Create a new `Ranged` value.
+    ///
+    /// Safety: the passed argument must be within `RANGE`
+    #[unstable(feature = "ranged_int", issue = "none")]
+    #[rustc_const_unstable(feature = "ranged_int", issue = "none")]
+    #[inline(always)]
+    pub const unsafe fn new_unchecked(i: T) -> Self {
+        Self(i)
+    }
+
+    /// Fetch the wrapped value.
+    #[unstable(feature = "ranged_int", issue = "none")]
+    #[rustc_const_unstable(feature = "ranged_int", issue = "none")]
+    #[inline(always)]
+    pub const fn get(self) -> T {
```

Comment:
Why do the `get` and `new_unchecked` implementations require `T: CheckInRange` when they don't seem to need it? And if it is needed, shouldn't the struct itself or at least the `Deref` implementation also have this requirement? Otherwise, you can easily bypass the requirement with `transmute` (for `new_unchecked`) and `deref` (for `get`), the latter of which would be the most concerning since it's safe and usually automatic.